### ISSUE TITLE
Add caching to aggregator responses

### ISSUE THREAD

**Issue by vlad-ignatov** [smart-on-fhir/cumulus-aggregator#132]
There are multiple caching strategies we can try. I'd propose we start with the simplest, which could look like this:

1. You get the modification date of the underlying s3 file
2. Make sure you send that date in a `Last-Modified` header
3. You **may** also get a `if-modified-since` header. If so, compare and:
    - `Last-Modified > if-modified-since` - exit with 304 status code without doing anything else.
    - `Last-Modified <= if-modified-since` - Continue with the queries as usual
    
Notes:
- If you send Last-Modified  you should get back if-modified-since  in the follow up requests. Otherwise it may not be present.
- The actual storage in this case happens in browsers.
- Browsers have mechanisms to clear the cache or to bypass it by force reloading (like Cmd+Shift+R on Mac)
- If this is not good enough (if the file modification timestamp is not the only variable we need to account for) we can switch to `ETag` caching which is very similar
- See https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Last-Modified and https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/If-Modified-Since

**Comment by dogversioning** [comment]
from vlad:

So, this is JS pseudo code but it should be easy to convert to Python:
// Requesting endpoint like {{baseUrl}}/data-packages

const TTL = 3600 // cache lifetime in seconds

// Create a hash from all the variables that would invalidate the cache
const ETag = crypto.createHash("sha1").update([
    requestedUrl,     // the full request URL, including the query string
    lastModifiedDate, // timestamp of last data update
    otherVars         // Other variables if needed
].join("-")).digest("hex");

// Always send these headers
response.setHeader('Cache-Control', `max-age=${TTL}, no-cache`)
response.setHeader('Vary', 'Origin, ETag')
response.setHeader('ETag', ETag)


const ifNoneMatchValue = request.headers['if-none-match']
if (ifNoneMatchValue && ifNoneMatchValue === ETag) {
    response.statusCode = 304
    response.end() // send empty 304 response
} 
else {
    response.statusCode = 200 // or any error code in case of error
    // Actual, slow, expensive job here...
    // process and respond with JSON 
}
How it works:

Basically, the client is stating one of these:
I don’t have anything locally. Just give me all the data packages
I do have data packages identified by this ETag. Please determine if that is up to date. If so, give me empty 304. Otherwise. give me the JSON.

- On any request
    - Create an ETag and send it to the client as header
    - Also instruct the client not to use cache older than TTL
- On first request
    - The client won’t provide if-none-match header so just build and return the JSON payload with status 200
- On follow up requests
    - The client should provide ETag in a `if-none-match` header. Compare that with the one you built
        - If ETags are the same, return empty 304 response
        - otherwise return the JSON payload with status 200
